scipy/special/_multiufuncs.py

```diff
@@ -0,0 +1,640 @@
+
+    Notes
+    -----
+    With respect to the associated Legendre polynomial :math:`P_{m}^{n}(\cos \phi)`,
+    the spherical Legendre polynomial is defined as
+
+    .. math::
+
+        \sqrt{\frac{(2 n + 1) (n - m)!}{4 \pi (n + m)!}} P_{n}^{m}(\cos \phi)
+
+    This is the same as the spherical harmonic :math:`Y_{m}^{n}(\theta, \phi)`
+    with :math:`\theta = 0`.
+    """, diff_n=0
+)
+
+
+@sph_legendre_p.override_key
+def _(diff_n):
+    diff_n = _nonneg_int_or_fail(diff_n, "diff_n", strict=False)
+    if not 0 <= diff_n <= 2:
+        raise ValueError(
+            "diff_n is currently only implemented for orders 0, 1, and 2,"
+            f" received: {diff_n}."
+        )
+    return diff_n
+
+
+sph_legendre_p_all = MultiUFunc(sph_legendre_p_all,
+    r"""sph_legendre_p_all(n, m, phi, *, diff_n=0)
+
+    Spherical Legendre polynomial of the first kind.
+
+    Parameters
+    ----------
+    n : array_like, int
+        Order (max) of the Legendre polynomial, must have ``n >= 0``.
+    m : array_like, int
+        Degree (max) of the Legendre polynomial, must have ``m >= 0``
+    phi : array_like, float
+        Input value.
+    diff_n : Optional[int]
+        A non-negative integer. Compute and return all derivatives up
+        to order ``diff_n``.
+
+    Returns
+    -------
+    p : ndarray or tuple[ndarray]
+        The spherical Legendre polynomial with ``diff_n`` derivatives,
+        each having shape (n + 1, 2 * m + 1, ...).
+
+    Notes
+    -----
+    With respect to the associated Legendre polynomial :math:`P_{m}^{n}(\cos \phi)`,
+    the spherical Legendre polynomial is defined as
+
+    .. math::
+
+        \sqrt{\frac{(2 n + 1) (n - m)!}{4 \pi (n + m)!}} P_{n}^{m}(\cos \phi)
+
+    This is the same as the spherical harmonic :math:`Y_{m}^{n}(\theta, \phi)`
+    with :math:`\theta = 0`.
+    """, diff_n=0
+)
+
+
+@sph_legendre_p_all.override_key
+def _(diff_n):
+    diff_n = _nonneg_int_or_fail(diff_n, "diff_n", strict=False)
+    if not 0 <= diff_n <= 2:
+        raise ValueError(
+            "diff_n is currently only implemented for orders 0, 1, and 2,"
+            f" received: {diff_n}."
+        )
+    return diff_n
+
+
+@sph_legendre_p_all.override_ufunc_default_kwargs
+def _(diff_n):
+    return {'axes': [()] + (diff_n + 1) * [(0, 1)]}
+
+
+@sph_legendre_p_all.override_resolve_out_shapes
+def _(n, m, z_shape, nout):
+    if ((not np.isscalar(n)) or (n < 0)):
+        raise ValueError("n must be a non-negative integer.")
+
+    return nout * ((n + 1, 2 * abs(m) + 1) + z_shape,)
+
+
+assoc_legendre_p = MultiUFunc(assoc_legendre_p,
+    r"""assoc_legendre_p(n, m, z, *, typ=2, norm=False, diff_n=0)
+
+    Associated Legendre polynomial of the first kind.
+
+    Parameters
+    ----------
+    n : array_like, int
+        Order of the Legendre polynomial, must have ``n >= 0``.
+    m : array_like, int
+        Degree of the Legendre polynomial.
+    z : array_like, float
+        Input value.
+    typ : Optional[int]
```

Comment:
+1 for a more descriptive name

Reply:
Ok, some history here. Previously there was an argument called `type` here. What I did so far was rename it to `typ` to prevent clashing. I actually have never used this argument. My online searching indicates nearly all the associated Legendre polynomial implementations I can find don't have this argument.

Of course, it's valuable, so we shouldn't remove it or anything. Mathematica _does_ have this argument and calls it "type". Maple also has an option for this and call is "LegendreCut" [2].

Is there some reason not to just call it `branch_cut` or `cut` like Maple?

[1] https://mathoverflow.net/questions/62358/legendre-function-types
[2] https://www.maplesoft.com/support/help/maple/view.aspx?path=Legendre

Reply:
I could accept just `branch_cut`. I suggested `branch_cut_type` for more continuity with the previous name.

Reply:
Have renamed it to `branch_cut`.